### ISSUE TITLE
learning rate decay and schedule

### ISSUE THREAD

**Issue by ccchang0111** [bojone/bert4keras#45]
請問可以解釋一下 learning_rate, weight_decay, lr_schedule之間的關係嗎？

假設

```python
num_train_steps = 100 
num_warmup_steps = 10

learning_rate = 0.01
weight_decay_rate = 0.05

lr_schedule = {
    num_warmup_steps : 0.99,
    num_train_steps: 0.01,
}

optimizer = extend_with_weight_decay(Adam)
optimizer = extend_with_piecewise_linear_lr(optimizer)
optimizer_params = {
        'learning_rate': learning_rate,
        'lr_schedule': lr_schedule,
        'weight_decay_rate': weight_decay_rate,
    }
optimizer = optimizer(**optimizer_params)
```

我的理解是總共train 100 steps,
0-10： 0 to (0.01\*0.99) (grow linearly)
10-100: 0.01 to (0.01\*0.01) (decay linearly)

weight_decay_rate又是用在那裡呢？

感謝

**Comment by bojone** [comment]
learning_rate是一个常数

lr_schedule是将学习率调节为分段线性的，比如lr_schedule={1000:2, 2000:0.5}就是说0～1000步内将学习率从0线性递增到2倍，然后1000～2000步内从2倍线性递减到0.5倍，2000步后保持0.5倍不变。

weight_decay_rate就跟AdamW的weight_decay一致，就是每次的更新量多了 (-weight_decay_rate * learning_rate * weight) 这部分。

优化器介绍： https://kexue.fm/archives/7094

**Comment by ccchang0111** [comment]
感謝解釋！
